extensions/vscode/lsp-client/src/extension.ts

```diff
@@ -51,12 +46,316 @@ export function activate(context: ExtensionContext) {
+			}
+		});
+	}
+}
+
 export function deactivate(): Thenable<void> | undefined {
 	if (!client) {
 		return undefined;
 	}
 	return client.stop();
 }
+
+function getWebviewOptions(): vscode.WebviewOptions {
+	return {
+		// Enable javascript in the webview
+		enableScripts: true,
+		// And restrict the webview to not loading any content.
+		localResourceRoots: []
+	};
+}
+
+/**
+ * Manages timeline webview panels
+ */
+class TimelinePanel {
+	/**
+	 * Track the current panel. Only allow a single panel to exist at a time.
+	 */
+	public static currentPanel: TimelinePanel | undefined;
+
+	public static readonly viewType = 'timeline';
+
+	private readonly _panel: vscode.WebviewPanel;
+	private _disposables: vscode.Disposable[] = [];
+
+	public static createOrShow() {
+		const column = vscode.window.activeTextEditor
+			? vscode.window.activeTextEditor.viewColumn
+			: undefined;
+
+		// If we already have a panel, show it.
+		if (TimelinePanel.currentPanel) {
+			TimelinePanel.currentPanel._panel.reveal(column);
+			return;
+		}
+
+		// Otherwise, create a new panel.
+		const panel = vscode.window.createWebviewPanel(
+			TimelinePanel.viewType,
+			'Timeline',
+			column || vscode.ViewColumn.One,
+			getWebviewOptions(),
+		);
+
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	public static revive(panel: vscode.WebviewPanel) {
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	private constructor(panel: vscode.WebviewPanel) {
+		this._panel = panel;
+
+		// Set the webview's initial html content
+		this._update();
+
+		// Listen for when the panel is disposed
+		// This happens when the user closes the panel or when the panel is closed programmatically
+		this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+		// Update the content based on view changes
+		this._panel.onDidChangeViewState(
+			e => {
+				if (this._panel.visible) {
+					this._update();
+				}
+			},
+			null,
+			this._disposables
+		);
+
+		// Handle messages from the webview
+		this._panel.webview.onDidReceiveMessage(
+			message => {
+				switch (message.command) {
+					case 'alert':
+						vscode.window.showErrorMessage(message.text);
+						return;
+				}
+			},
+			null,
+			this._disposables
+		);
+	}
+
+	private _getDB() {
+		return new sqlite3.Database(__dirname + '/../../../../wake.db', sqlite3.OPEN_READONLY, (err) => {
+			if (err) {
+				return console.error(err.message);
+			}
+		});
+	}
+
+	public refresh() {
+		let db = this._getDB();
+		this._queryDB(db, (jobsJson: JobReflection[], accessesJson: { type: number, job: number }[]) => {
+			let message = {
+				jobReflections: jobsJson,
+				fileAccesses: accessesJson
+			}
+			// Send a message to the webview.
+			this._panel.webview.postMessage(message);
+			db.close();
+		});
+	}
+
+	public dispose() {
+		TimelinePanel.currentPanel = undefined;
+
+		// Clean up our resources
+		this._panel.dispose();
+
+		while (this._disposables.length) {
```

Comment:
Nit: I think it would be more idiomatic to just normally loop over this and then truncate the array afterwards. `while` loops are more powerful, but force readers to have to read more carefully to understand what's happening in the loop body. `for-of` loops are weaker and immediately communicate a few things to the reader:
- That you are operating on an array
- That you are going to access each element of the array once
- That you are going to iterate over the array in forwards order

Reply:
Done